File: gajae-company/scripts/logger.ts
import { initializeApp, cert } from 'firebase-admin/app';
import { getFirestore, FieldValue } from 'firebase-admin/firestore';
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';

/**
 * [가재 컴퍼니] Standard Intelligence Logger (v2.0)
 * 의도: 마크다운 구조를 분석하여 Firestore의 정규화된 필드에 지능을 박제함.
 * 정책: 'commands'와 'meetings' 컬렉션을 분리하고 구조화된 데이터(JSON)로 저장.
 */

const SERVICE_ACCOUNT_PATH = '/Users/openclaw-kong/.openclaw/workspace/firebase-service-account.json';

if (!fs.existsSync(SERVICE_ACCOUNT_PATH)) {
    console.error("❌ Error: Firebase Service Account key not found.");
    process.exit(1);
}

const serviceAccount = JSON.parse(fs.readFileSync(SERVICE_ACCOUNT_PATH, 'utf8'));

initializeApp({
    credential: cert(serviceAccount),
    projectId: 'gajae-company-bip'
});

const db = getFirestore();

interface FivefoldProtocol {
    intent: string;
    psychology: string;
    thought: string;
    action: string;
    response: string;
}

function parseMarkdown(content: string, type: 'command' | 'meeting') {
    const lines = content.split('\n');
    const result: any = { rawContent: content };

    // Common Metadata Parsing
    const timeMatch = content.match(/- \*\*일시\*\*: (?:.*?) (?:(\d{2}:\d{2}:\d{2})|(\d{2}:\d{2}))/);
    result.time = timeMatch ? (timeMatch[1] || timeMatch[2]) : "00:00";

    // Fivefold Protocol Parsing
    const extractSection = (regex: RegExp) => {
        const match = content.match(regex);
        return match ? match[1].trim() : "";
    };

    const protocol: FivefoldProtocol = {
        intent: extractSection(/\d\. \*\*의도 \(Intention\)\*\*: ([\s\S]*?)(?=\n\d\.|$)/),
        psychology: extractSection(/\d\. \*\*심리 \(Psychology\)\*\*: ([\s\S]*?)(?=\n\d\.|$)/),
        thought: extractSection(/\d\. \*\*생각 \(Thought\)\*\*: ([\s\S]*?)(?=\n\d\.|$)/),
        action: extractSection(/\d\. \*\*행동 \(Action\)\*\*: ([\s\S]*?)(?=\n\d\.|$)/),
        response: extractSection(/\d\. \*\*답변 \(Response\)\*\*: ([\s\S]*?)(?=\n\d\.|$)/)
    };

    if (type === 'command') {
        result.instruction = extractSection(/## 📜 지시 내용 \(Command\)\n([\s\S]*?)(?=\n---|$)/);
        result.execution = protocol;
    } else {
        const hostMatch = content.match(/- \*\*주관\*\*: (.*?)\n/);
        result.host = hostMatch ? hostMatch[1].trim() : "";
        const participantsMatch = content.match(/- \*\*참석\*\*: (.*?)\n/);
        result.participants = participantsMatch ? participantsMatch[1].split(',').map(p => p.trim()) : [];
        result.details = protocol;
    }

    return result;
}

async function logToFirestore(type: 'command' | 'meeting', title: string, author: string, content: string) {
    const now = new Date();
    const dateStr = now.toISOString().split('T')[0].replace(/-/g, '');
    const docId = crypto.randomUUID();
    
    const parsedData = parseMarkdown(content, type);
    const collectionName = type === 'command' ? 'commands' : 'meetings';

    const finalData = {
        id: docId,
        date: dateStr,
        title,
        author,
        ...parsedData,
        createdAt: FieldValue.serverTimestamp()
    };

    try {
        await db.collection(collectionName).doc(docId).set(finalData);
        console.log(`✅ Log structured and persisted to [${collectionName}]: ${docId}`);
        return docId;
    } catch (error) {
        console.error("❌ Failed to log to Firestore:", error);
        throw error;
    }
}

const args = process.argv.slice(2);
if (args.length >= 4) {
    const [type, title, author, content] = args;
    logToFirestore(type as any, title, author, content).catch(() => process.exit(1));
} else {
    console.log("Usage: npx tsx scripts/logger.ts <command|meeting> <title> <author> <content>");
}

File: gajae-company/scripts/logger.py
import firebase_admin
from firebase_admin import credentials, firestore
import re
import uuid
import sys
import os
from datetime import datetime

# [가재 컴퍼니] Standard Intelligence Logger (Python v2.0)
# 의도: 마크다운 구조를 정밀 분석하여 Firestore의 정규화된 필드에 지능을 박제함.

SERVICE_ACCOUNT_PATH = '/Users/openclaw-kong/.openclaw/workspace/firebase-service-account.json'

def init_firebase():
    if not firebase_admin._apps:
        cred = credentials.Certificate(SERVICE_ACCOUNT_PATH)
        firebase_admin.initialize_app(cred)

def parse_markdown(content, log_type):
    result = {'rawContent': content}
    
    # Extract Time
    time_match = re.search(r'-\s\*\*일시\*\*:\s(?:.*?)\s(?:(\d{2}:\d{2}:\d{2})|(\d{2}:\d{2}))', content)
    result['time'] = time_match.group(1) or time_match.group(2) if time_match else "00:00"

    # Fivefold Protocol Sections
    protocol = {}
    sections = [
        ('intent', r'1\.\s\*\*의도\s\(Intention\)\*\*:\s([\s\S]*?)(?=\n\d\.|$)'),
        ('psychology', r'2\.\s\*\*심리\s\(Psychology\)\*\*:\s([\s\S]*?)(?=\n\d\.|$)'),
        ('thought', r'3\.\s\*\*생각\s\(Thought\)\*\*:\s([\s\S]*?)(?=\n\d\.|$)'),
        ('action', r'4\.\s\*\*행동\s\(Action\)\*\*:\s([\s\S]*?)(?=\n\d\.|$)'),
        ('response', r'5\.\s\*\*답변\s\(Response\)\*\*:\s([\s\S]*?)(?=\n\d\.|$)'),
    ]

    for key, pattern in sections:
        match = re.search(pattern, content)
        protocol[key] = match.group(1).strip() if match else ""

    if log_type == 'command':
        instr_match = re.search(r'##\s📜\s지시\s내용\s\(Command\)\n([\s\S]*?)(?=\n---|$)', content)
        result['instruction'] = instr_match.group(1).strip() if instr_match else ""
        result['execution'] = protocol
    else:
        host_match = re.search(r'-\s\*\*주관\*\*:\s(.*?)\n', content)
        result['host'] = host_match.group(1).strip() if host_match else ""
        part_match = re.search(r'-\s\*\*참석\*\*:\s(.*?)\n', content)
        result['participants'] = [p.strip() for p in part_match.group(1).split(',')] if part_match else []
        result['details'] = protocol

    return result

def log_to_firestore(log_type, title, author, content):
    init_firebase()
    db = firestore.client()
    
    parsed = parse_markdown(content, log_type)
    collection = 'commands' if log_type == 'command' else 'meetings'
    
    doc_id = str(uuid.uuid4())
    data = {
        'id': doc_id,
        'date': datetime.now().strftime("%Y%m%d"),
        'title': title,
        'author': author,
        'createdAt': firestore.SERVER_TIMESTAMP,
        **parsed
    }

    db.collection(collection).document(doc_id).set(data)
    print(f"✅ Log structured and persisted to [{collection}]: {doc_id}")

if __name__ == "__main__":
    if len(sys.argv) >= 5:
        log_to_firestore(sys.argv[1], sys.argv[2], sys.argv[3], sys.argv[4])
    else:
        print("Usage: python3 scripts/logger.py <command|meeting> <title> <author> <content>")
